src/tls/ffi.c

```diff
@@ -0,0 +1,428 @@
+    if (!g_mitls_FFI_##x) { \
+        printf("Failed to bind to Caml callback MITLS_FFI_" # x "\n"); \
+        return 0; \
+    }
+ MITLS_FFI_LIST  
+ #undef MITLS_FFI_ENTRY
+    
+    return 1; // success
+}
+
+void FFI_mitls_cleanup(void)
+{
+#define MITLS_FFI_ENTRY(x) \
+    g_mitls_FFI_##x = NULL;
+ MITLS_FFI_LIST  
+ #undef MITLS_FFI_ENTRY
+}
+
+char * read_stdio_file(int fd)
+{
+    struct stat st;
+    char *data;
+    
+    fstat(fd, &st);
+    // If no data was written, or a huge amount of data was written, ignore the file.
+    if (st.st_size == 0 || st.st_size > 10*1024*1024) {
+        return NULL;
+    }
+    
+    lseek(fd, SEEK_SET, 0);
+    data = malloc(st.st_size + 1); // The +1 is safe from integer overflow due to the above size check
+    if (!data) {
+        return NULL;
+    }
+    read(fd, data, st.st_size);
+    data[st.st_size] = '\0'; // Null-terminate into a C string
+    
+    return data;
+}
+
+void restore_stdio(/* in */ mitls_redirect *r, /* out */ char **outmsg, /* out */ char **errmsg)
+{
+#if REDIRECT_STDIO
+    char *msg;
+    int result;
+    
+    *outmsg = NULL;
+    *errmsg = NULL;
+ 
+    if (r->fd_out_new == -1) {
+        return;
+    }
+    
+    fflush(stdout);
+    fflush(stderr);
+    
+    // Restore stdout/stderr back to their original files
+    dup2(r->fd_out_old, STDOUT_FILENO);
+    dup2(r->fd_err_old, STDERR_FILENO);
+    
+    msg = read_stdio_file(r->fd_out_new);
+    if (msg) {
+        *outmsg = msg;
+    }
+    close(r->fd_out_new);
+    result = unlink(r->out_name);
+    if (result != 0) {
+        fprintf(stderr, "unlink %s failed errno=%d ", r->out_name, errno); perror("failed");
+    }
+    free(r->out_name);
+    
+    msg = read_stdio_file(r->fd_err_new);
+    if (msg) {
+        *errmsg = msg;
+    }
+    close(r->fd_err_new);
+    result = unlink(r->err_name);
+    if (result != 0) {
+        fprintf(stderr, "unlink %s failed errno=%d ", r->out_name, errno); perror("failed");
+    }
+    free(r->err_name);
+#else
+    r->fd_out_new = -1;
+    *outmsg = NULL;
+    *errmsg = NULL;
+#endif
+}
+
+
+void redirect_stdio(/* out */ mitls_redirect *r)
+{
+#if REDIRECT_STDIO
+    int fdout;
+    int fderr;
+    static const char template[] = "/mitlscurlXXXXXX";
+    char *tmp;
+    char *out_name;
+    char *err_name;
+    size_t tmp_len;
+    
+    // Assume failure
+    r->fd_out_new = -1;
+    r->fd_err_new = -1;
+    
+    fflush(stdout);
+    fflush(stderr);
+    
+    tmp = getenv("TMP");
+    if (!tmp) {
+        tmp = getenv("TEMP");
+        if (!tmp) {
+            tmp = ".";
+        }
+    }
+    
+    tmp_len = strlen(tmp);
+    out_name = (char*)malloc(tmp_len + sizeof(template));
+    if (!out_name) {
+        return;
+    }
+    strcpy(out_name, tmp);
+    strcpy(out_name+tmp_len, template);
+    
+    err_name = strdup(out_name);
+    if (!err_name) {
+        free(out_name);
+        return;
+    }
+
+    // Create temp files
+    fdout = mkstemp(out_name);
```

Comment:
oh ok, missed that -- you're insisting that the file be located in the TMP or TEMP directory, then providing the template to mkstemp